debian/exe.1
'\" t
.\"     Title: INTEF-EXE
.\"    Author: Juan Rafael Fern\('andez <jrfern@telefonica.net>
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 03/28/2021
.\"    Manual: eXe User Manual
.\"    Source: exe
.\"  Language: English
.\"
.TH "INTEF\-EXE" "1" "03/28/2021" "exe" "eXe User Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
exe \- The EXtremely Easy to use eLearning authoring tool
.SH "SYNOPSIS"
.HP \w'\fBexe\fR\ 'u
\fBexe\fR [{\fB\-h\fR\ |\ \fB\-\-help\fR} | {\fB\-V\fR\ |\ \fB\-\-version\fR} | \fB\-\-standalone\fR  | \fB\-\-portable\fR ]
.SH "DESCRIPTION"
.PP
This manual page documents briefly the
\fBexe\fR
command\&.
.PP
\fBexe\fR
is a GPL2+ free/libre tool for the educational community to create and publish open educational resources\&.
.SH "OPTIONS"
.PP
The program follows the usual GNU command line syntax, with long options starting with two dashes (`\-\*(Aq)\&. A summary of options is included below\&.

.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Display this help and exit\&.
.RE
.PP
\fB\-V\fR, \fB\-\-version\fR
.RS 4
Print version information and exit\&.
.RE
.PP
\fB\-\-standalone\fR
.RS 4
Run totally from the current directory (doesn\*(Aqt use the system libraries)\&.
.RE
.PP
\fB\-\-portable\fR
.RS 4
Run in portable mode\&.
.RE
.SH "FILES"
.PP
${HOME}/\&.exe/exe\&.conf
.RS 4
The per\-user configuration file to control the behaviour of
exe\&.
.RE
.SH "BUGS"
.PP
The upstreams
BTS
can be found at
\m[blue]\fB\%https://github.com/exelearning/iteexe/issues\fR\m[]\&.
.SH "SEE ALSO"
.PP
\fBexe_do\fR(1)
.PP
The programs are documented fully by
eXeLearning Manual
available at
\m[blue]\fB\%http://exelearning.net/html_manual/exe20/\fR\m[]\&.
.SH "AUTHOR"
.PP
\fBJuan Rafael Fern\('andez\fR <\&jrfern@telefonica\&.net\&>
.RS 4
Wrote this manpage for the Debian system\&.
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2015 Juan Rafael Fern\('andez
.br
.PP
This manual page was written for the Debian system (and may be used by others)\&.
.PP
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 3 or any later version published by the Free Software Foundation\&.
.PP
On Debian systems, the complete text of the GNU General Public License can be found in
/usr/share/common\-licenses/GPL\&.
.sp


NEWS
* Add max score reporting to LMS for SCORM Quiz
* Add username to startup lock file to permit multiple users on same machine
* Fix startup with Firefox 3.x tabs hiding interface (Ivo Benner)
* Fix image gallery to allow captions containing apostrophes
* Fix field IDs so that adding to a previously saved elp is consistent
* Fix Common Cartridge manifest when a header image is used
* Fix OS X and Linux configDir Firefox profile on filesystem without colon
* Fix display of translated help for Multi-Select iDevice editing
* Workaround AVG anti-virus adding itself to eXe Firefox on Windows
* Make startup lock file parsing more robust
* Only include a copy of GNU FDL if that is the license selected
* Re-enable loading of user extended iDevices from config directory
* Translation updates

Version 1.04 (r3532) 2008-05-09
* Add offline snapshot of manual, release notes, and tutorial to Help menu
* Add undo button to discard changes after editing
* Add automatic auto_top anchors to facilitate internal linking to each node
* Add ability to translate tooltips
* Fix handling of multi-select questions with 10 or more options
* Fix tip display at right edge of windows
* Fix Common Cartridge resource references
* Fix package property display with Firefox 3.0 (no properties translations)
* Improve Common Cartridge metadata formation
* Improve Wiki Article iDevice handling of pages without content div
* Improve Firefox version detection for 3.0b series (Linux)
* Remove CSS letter-spacing to improve rendering with some fonts (Khmer)
* Prototype ability to load eXe-authored Common Cartridges
* Automate building Debian changelog and versioning
* Update to BeautifulSoup 3.0.6 (character coding detection issues)
* Translation updates and additions

Version 1.03 (r3373) 2008-01-21
* Add linking across pages within a package (only for web export)
* Add note that SCORM export previous/next requires non-standard extension
* Add prototype IMS Common Cartridge export (no quizzing)
* Add Rel-License microformat to license urls in exported content
* Make importing old packages more robust (and remove unused resources)
* Fix merging packages that contain resources with same name as main package
* Fix upgrading of multichoice questions
* Fix escaping of ampersands in rich text fields, allowing HTML embedding
* Fix embedding of images and media with '#' in name
* Fix SCORM Quiz pass rate to signal pass for scores of exactly the rate
* Update to BeautifulSoup 3.0.5
* Translation updates

Version 1.02 (r3303) 2007-10-01
* Add ability to type resource name into url box (without browse button)
* Add CSS divs around nodes in single page and print exports for pagination
* Add detailed uninstall name, links, icon for Windows
* Make reading very old eXe packages more robust if corruption encountered
* Fix XHTML for div wrapper of Java Applet iDevice export
* Add ID, TH translations

Version 1.01 (r3265) 2007-09-05
* Add embedded MP3 player as a rich text editor media type
* Add FLV player as rich text editor media type
* Add DIV around each iDevice on web/ims/scorm export
* Add disabling SUBMIT ANSWERS button in SCORM Quiz after first submission
* Fix web folder export to convert filenames to safe ASCII characters
* Fix iPod Notes export to write UTF-8 index (non-ASCII node names)
* Fix XHTML generated for SCORM Quiz
* Fix XHTML generated for True-False question feedback
* Fix reading of exe.conf files that have UTF-8 "byte order mark"
* Fix insertion of non-ASCII package names in Recent Projects list
* Change license footer text so translators can ignore trailing space
* Change RPM spec file to work better with openSUSE
* Remove deprecated dublincore.xml metadata file from SCORM 1.2 packages
* Updated DE,FR,IT,PL translations

Version 1.00 (r3219) 2007-08-17
* Fix styles to use relative line height rather than absolute
* Fix splash progress bar to indicate 100%
* Fix iDevice deletion issues including read-only resources
* Fix display of embedded image borders
* Fix header text display in single page export and printing
* Make merging packages more robust
* Make default IMS/SCORM Metadata title description from package values
* Change export to folder to not change UTF-8 filenames
* Update About box

Version 0.99 (r3186) 2007-08-10
* Add splash screen for when launching eXe by double-clicking .elp
* Add warning screen for trying to launch multiple copies of eXe
* Add warning for trying to run an older version of eXe with newer data
* Add Catalan Wikipedia
* Add file browser to rich text editor link popup for including files
* Fix re-enable Flash autoplay attribute
* Fix multi-select option feedback
* Fix SCORM Quiz to allow deleting individual questions
* Change text export to decode HTML entities into UTF-8 string
* Remove media plugin detection allowing authoring without media plugin
* Allow specifying VCARD type in metadata fields for SCORM export
* Allow non-ASCII characters in embedded media filenames
* Slightly reduce eXe package size by not storing multiple copies of text
* Updated German, Japanese, and Chinese translations

Version 0.98 (r3136) 2007-08-03
* Add ability to insert MimeTeX math images in any field with TinyMCE
* Add message to RSS iDevice for when feed fails
* Add labels to items added in the iDevice editor edit display
* Add feedback display for each option of multi-select iDevice (not score)
* Add font family, background colors, advanced horizontal rule to TinyMCE
* Fix Image Gallery popup to start at images beyond first row
* Fix Image Gallery to handle unicode captions and previous/next
* Fix SCORM Quiz scoring to send normalized 0..100 score to LMS
* Fix SCORM 1.2 export that was missing ims_xml schema
* Fix display of package header image by adding a Show Image button
* Change embedded media to default to auto-fit and not autoplay
* Attempt to decode HTML character entities on text export
* Start using IMS LRM metadata for SCORM 1.2 exports

exe/jsui/scripts/exe.js
// ===========================================================================
// eXe
// Copyright 2012, Pedro Peña Pérez, Open Phoenix IT
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//===========================================================================

if (!window.translations)
    translations = {};

function _(msg) {
    return translations[msg] || msg;
}

Ext.Loader.setConfig({
    enabled: true,
    paths: { 'Ext.ux': 'jsui/extjs/examples/ux' }
});

var conf = {
    onRender: function() {
        var me = this;
        
        me.callParent(arguments);
        if (me.tooltip) {
            Ext.tip.QuickTipManager.register(Ext.apply({
                target: me.el,
                text: me.tooltip
            }));
        }
    }
};

//Call authoring page when zindex is modified and consider problematic plugins with no zindex support
Ext.override(Ext.WindowManager, {
    bringToFront: function(comp) {
        var me = this, authoringPanel = Ext.ComponentQuery.query('#authoring')[0];

        me.callParent(arguments);
        if (authoringPanel.isVisible()) {
            var authoring = authoringPanel.getWin();
	        if (authoring && authoring.hideObjectTags)
	            authoring.hideObjectTags();
        }
    },
    onComponentHide: function(comp) {
        var me = this, authoringPanel = Ext.ComponentQuery.query('#authoring')[0];

        me.callParent(arguments);
        if (authoringPanel.isVisible()) {
            if (!this.getActive()) {
		        var authoring = authoringPanel.getWin();
		        if (authoring && authoring.showObjectTags)
		            authoring.showObjectTags();
            }
        }
    },
    _hideModalMask: function() {
        var me = this, authoringPanel = Ext.ComponentQuery.query('#authoring')[0];

        me.callParent(arguments);
		if (authoringPanel.isVisible()) {
            if (!this.getActive()) {
		        var authoring = authoringPanel.getWin();
		        if (authoring && authoring.showObjectTags)
		            authoring.showObjectTags();
            }
        }
    }
});
Ext.override(Ext.form.field.Text, conf);
Ext.override(Ext.form.field.TextArea, conf);
Ext.override(Ext.form.field.Checkbox, conf);
Ext.override(Ext.form.field.Hidden, conf);

Ext.override(Ext.menu.Menu,{
      onMouseOver: function(e) {
            var me = this,
                  fromEl = e.getRelatedTarget(),
                  mouseEnter = !me.el.contains(fromEl),
                  item = me.getItemFromEvent(e),
                  parentMenu = me.parentMenu,
                  ownerCmp = me.ownerCmp;

            if (mouseEnter && parentMenu) {
                  //original
                  //parentMenu.setActiveItem(ownerCmp);
                  //ownerCmp.cancelDeferHide();
                  //parentMenu.mouseMonitor.mouseenter();
                  //end original

                  //fix
                  if(ownerCmp){
                        parentMenu.setActiveItem(ownerCmp);
                        ownerCmp.cancelDeferHide();
                  }
                  setTimeout(parentMenu.mouseMonitor.mouseenter,5);
                  //end fix
            }

            if (me.disabled) {
                  return;
            }

            if (item && !item.activated) {
                  me.setActiveItem(item);
                  if (item.activated && item.expandMenu) {
                        item.expandMenu();
                  }
            }
            if (mouseEnter) {
                  me.fireEvent('mouseenter', me, e);
            }
            me.fireEvent('mouseover', me, item, e);
      }
});

Ext.application({
    name: 'eXe',

    stores: [
        'OutlineXmlTreeStore',
        'IdeviceXmlStore',
        'filepicker.DirectoryTree',
        'filepicker.File'
    ],

    models: [
    	'filepicker.File'
    ],
    
    controllers: [
        'eXeViewport',
    	'Idevice',
        'MainTab',
    	'Outline',
        'Toolbar',
        'StyleManager',
        'TemplateManager',
    	'filepicker.Directory',
    	'filepicker.File'
    ],
    
    getMaxHeight: function(height) {
        var vheight = Ext.ComponentQuery.query('#eXeViewport')[0].getHeight();

        return height >= vheight? vheight : height;
    },

    quitWarningEnabled: true,

    reload: function() {
        var authoring = Ext.ComponentQuery.query('#authoring')[0].getWin();
        if (authoring && authoring.submitLink) {
        	var outlineTreePanel = eXe.app.getController("Outline").getOutlineTreePanel(),
            	selected = outlineTreePanel.getSelectionModel().getSelection();
        	eXe.app.quitWarningEnabled = false;
        	eXe.app.on({
        		'authoringLoaded': function() {
        			nevow_clientToServerEvent('reload');
        		}
        	});
	        authoring.submitLink("changeNode", selected !== 0? selected[0].data.id : '0');
        }
    },

    gotoUrl: function(location) {
        eXe.app.quitWarningEnabled = false;
        if (location == undefined)
            location = window.top.location.pathname;
        nevow_closeLive('window.top.location = "' + location + '";');
    },
    
    alert: function(msg){
        Ext.Msg.alert(_("Warning"), msg);
    },
    
    notifications: {
        savedPackage : function(filePath) {
            Ext.Msg.close();
            eXe.controller.eXeViewport.prototype.eXeNotificationStatus(_("Success"),filePath);
            setTimeout(function(){
                Ext.ComponentQuery.query("#eXeNotification")[0].hide();
            },3000);
        }
    },
    
    showLoadError: function() {
    	if (eXe.app.config.loadErrors.length > 0) {
    		Ext.Msg.alert(_('Load Error'), eXe.app.config.loadErrors.pop(), eXe.app.showLoadError);
    	}
    	else
    		eXe.app.afterShowLoadErrors();
    },
    
    showNewVersionWarning: function(){
        // Show a warning message if a new version is available
        if (navigator.onLine && eXe.app.config.showNewVersionWarning && typeof(eXe.app.config.release)=='string') {
            function openNewVersionWarning(no){
                
                var latest = no;
                var current = eXe.app.config.release;
                
                if (current=="" || latest==current) return; // No release number for testing packages
                
                latest = latest.replace(/\./g,"");
                current = current.replace(/\./g,"");
				// Remove ~alfa or ~beta
				current = current.split("~");
				current = current[0];

                while (latest.length>current.length) current += '0';
                while (current.length>latest.length) latest += '0';                
                
                current = parseFloat(current);
                latest = parseFloat(latest);
                
                if (latest>current) {
                    var msg = _('A new version of eXeLearning (%) is available. Would you like to download it now?');
                    msg = msg.replace("%",no);
                    Ext.Msg.show({
                        title: _('Warning!'),
                        msg: msg,
                        scope: this,
                        modal: true,
                        buttons: Ext.Msg.YESNO,
                        fn: function(button) {
                            if (button == "yes") {
                                window.open("http://exelearning.net/");
                            }
                        }
                    });
                }                
            }
            Ext.data.JsonP.request({
                url: 'https://api.github.com/repos/exelearning/iteexe/releases/latest',
                crossDomain: true,
                type: "GET",
                dataType: "json",
                callbackKey: 'callback',
                scope: this,
                callback: function (response, value, request) {
                    if (typeof(value)!='undefined' && typeof(value.data)!='undefined' && typeof(value.data.name)=='string') {
                        openNewVersionWarning(value.data.name);
                    }
                }
            });
        }
    },
    
    launch: function() {
        Ext.QuickTips.init();

        try {
            Ext.state.Manager.setProvider(new Ext.state.LocalStorageProvider());
        }
        catch (err) {
            console.log('Local Storage not supported');
            Ext.state.Manager.setProvider(new Ext.state.CookieProvider({expires: null}));
        }
        
        eXe.app = this;
        
        eXe.app.config = config;
        
        Ext.state.Manager.set('filepicker-currentDir', eXe.app.config.lastDir);

        window.onbeforeunload = function() {
            if (eXe.app.quitWarningEnabled)
                return _("If you leave this page eXe application continues to run." +
                        " Please use the menu File->Quit if you really want to exit the application.");
        };
		/*
		window.onunload = function() {
            nevow_clientToServerEvent('quit', '', '');
        };
		*/

        langsStore.sort(function(a, b) {
            return a[1].localeCompare(b[1]);
        });

        if (Ext.isGecko || Ext.isSafari)
        	window.addEventListener('keydown', function(e) {(e.keyCode == 27 && e.preventDefault())});

        var cmp1 = Ext.create('eXe.view.ui.eXeViewport', {
            renderTo: Ext.getBody()
        });
        cmp1.show();

        setTimeout(function(){
		    Ext.get('loading').hide();
		    Ext.get('loading-mask').fadeOut();
		  }, 250);
        
        if (!eXe.app.config.showIdevicesGrouped) {
        	var panel = Ext.ComponentQuery.query('#idevice_panel')[0],
        		button = panel.down('button');
        	
        	panel.view.features[0].disable();
        	button.setText(_('Group iDevices'));
        }

        eXe.app.afterShowLoadErrors = function() {
        	if (eXe.app.config.showPreferences) eXe.app.getController('Toolbar').toolsPreferences(true); // true to open the new version warning
            else eXe.app.showNewVersionWarning();
        };

        eXe.app.showLoadError();
    },
    
    createLeftPanelToggler : function(isLoadEvent){
        if (isLoadEvent) {
            var a = document.getElementById("hide_exe_leftpanel");
            if (a && a.offsetParent != null) return; // The left panel is visible (do not create the link)
        }
        var iframe = document.getElementsByTagName('iframe');
        if (iframe.length==1) {
            iframe = iframe[0];
            var doc = iframe.contentWindow.document;
            lnk = doc.getElementById('show_exe_leftpanel');
            if (!lnk) {
                var l = document.createElement("a");
                l.href = "#";
                l.id = "show_exe_leftpanel";
                l.title = _("Show panel");
                l.style.position = "fixed";
                l.style.top = "5px";
                l.style.left = "0";
                l.style.outline = "none";
                l.onclick = function(){
                    var panel = top.Ext.ComponentQuery.query('#exe_leftpanel')[0];
                    panel.show();
                    this.style.display = 'none';
                    return false;
                }
                l.innerHTML = '<img width="35" height="37" src="/images/show-left-panel.png" alt="'+_("Show panel")+'" />';
                doc.body.appendChild(l);
            } else {
                lnk.style.display = 'inline';
            }
        }
    },
    
    createEmptyPageInstructions : function(){
        var a = document.getElementById("hide_exe_leftpanel");
        if (a && a.offsetParent != null) { // The left panel is visible
            if (!Ext.util.Cookies.get('eXeShowContentInstructions')) {
                var iframe = document.getElementsByTagName('iframe');
                if (iframe.length==1) {
                    var iframe = iframe[0];
                    var doc = iframe.contentWindow.document;                    
                    if (doc && doc.getElementsByClassName) {
                        var iDevices = doc.getElementsByClassName("iDevice");
                        if (iDevices.length==0 && !doc.getElementById("activeIdevice")) {
                            var div = doc.getElementById("main");
                            if (div) {
                                div.style.minHeight = "400px";
                                var msg = _('Click on the elements of the left panel to add content.');
                                    msg += ' ' + '<a href="#" id="emptyPageInstructionsLink"><span class="sr-av">' + _("Don't show this warning again") + ' </span>×</a>';
                                div.innerHTML += '<p class="exe-block-info" id="emptyPageInstructions">'+msg+'</p>';
                                var a = doc.getElementById("emptyPageInstructionsLink");
                                if (a) {
                                    a.onclick = function(){
                                        var div = doc.getElementById("emptyPageInstructions");
                                        if (div) div.style.display = "none";
                                        Ext.util.Cookies.set('eXeShowContentInstructions', 'no');
                                        return false;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }        
    },
	
	// Add a class to the empty non-emphasized iDevices so the user can see the buttons to edit them
	// It also adds the exe-advanced CSS class to the body tag if needed
	checkIdevicesVisibility : function(){
		var iframe = document.getElementsByTagName('iframe');
		if (iframe.length==1) {
			var iframe = iframe[0];
			var doc = iframe.contentWindow.document;                    
			if (doc && doc.getElementsByClassName) {
				if (document.body.className.indexOf(" exe-advanced")!=-1 && doc.body.className.indexOf(" exe-advanced")==-1) doc.body.className += " exe-advanced";
				var iDevices = doc.getElementsByClassName("iDevice");
				if (iDevices.length>0) {
					for (var i=0;i<iDevices.length;i++) {
						var e = iDevices[i];
						if (e.className.indexOf(" emphasis0")!=-1) {
							var iDeviceContent = e.getElementsByClassName("iDevice_content");
							if (iDeviceContent.length>0) {
								iDeviceContent = iDeviceContent[0];
								var html = iDeviceContent.innerHTML;
									html = html.replace(/ /g,"");
									html = html.replace(/(?:\r\n|\r|\n)/g, "");
								if (html=="") e.className += " emphasis0-empty";
							}
						}
					}
				}
			}
		}
	},
	
	// Check if the advanced mode is on or not
	checkAdvancedModePreference : function(){
		if (typeof(advancedModePreferenceValue)=='undefined') advancedModePreferenceValue = 0;
		if (advancedModePreferenceValue==1) Ext.getCmp("advanced_toggler").setValue(true);
		else Ext.getCmp("advanced_toggler").setValue(false);
		eXe.app.getController("Toolbar").eXeUIversionSetStatus(advancedModePreferenceValue);
	},

    appFolder: "jsui/app"

});

